Model Registry Stack Component + MLFlow integration (#1309)
zenml-io/zenml: docs/book/component-gallery/model-registries/model-registries.md

---
description: How to track and manage ML models
---

Model registries are centralized storage solutions for managing and tracking
machine learning models across various stages of development and deployment.
They help track the different versions and configurations of each model and
enable reproducibility. By storing metadata such as version, configuration, and
metrics, model registries help streamline the management of trained models. In

> ZenML, model registries are Stack Components that allow for easy retrieval,

Reviewer: ```suggestion
ZenML, model registries are Stack Components that allow for the easy retrieval,
```

loading, and deployment of trained models. They also provide information on the
pipeline in which the model was trained and how to reproduce it.

## Model Registry Concepts and Terminology

> To handle and manage the concepts of model groups, versions, and stages, in a 
> consistent way regardless of the underlying registry tool or platform being used.
> ZenML provides a unified abstraction for model registries by defining the
> following concepts:

Reviewer: ```suggestion
ZenML provides a unified abstraction for model registries through which 
it is possible to handle and manage the concepts of model groups, versions, 
and stages in a consistent manner regardless of the underlying registry tool or platform being used.
The following following concepts are useful to be aware of for this abstraction:
```

> * **ModelRegistration**: ModelRegistration: A logical grouping of models that

Reviewer: ```suggestion
* **ModelRegistration**: A logical grouping of models that
```

    by the user or automatically created by the model registry when a new model
    is logged.

* **ModelVersion**: A specific version of a model identified by a unique version
    number or string. It holds information about the model, such as its name,
    description, tags, and metrics, and a reference to the model artifact logged
    to the model registry. In ZenML, it also holds a reference to the pipeline
    name, pipeline run ID, and step name. Each model version is associated with
    a model registration.

* **ModelVersionStage**: A model version stage is a state that a model version
    can be in. It can be one of the following: `None`, `Staging`, `Production`,
    `Archived`. The model version stage is used to track the
    lifecycle of a model version. For example, a model version can be in the
    `Staging` stage while it is being tested and then moved to the `Production`
    stage once it is ready for deployment.

## When to use it

ZenML provides a built-in mechanism for storing and versioning pipeline artifacts
through its mandatory Artifact Store. While this is a powerful way to manage

> artifacts programmatically, it can be challenging to work with without a visual

Reviewer: ```suggestion
artifacts programmatically, it can be challenging to use without a visual
```

interface.

Model registries, on the other hand, offer a visual way to manage and track
model metadata, particularly when using a remote orchestrator. They make it
easy to retrieve and load models from storage, thanks to built-in integrations.
A model registry is an excellent choice for interacting with all the models in
your pipeline and managing their state in a centralized way.

Using a model registry in your stack is particularly useful if you want to
interact with all of the logged models in your pipeline, or if you need to
manage the state of your models in a centralized way and make it easy to
retrieve, load, and deploy these models.

> ## How they model registries slot into the ZenML stack

Reviewer: ```suggestion
## How model registries fit into the ZenML stack
```

> Here is an architecture diagram that shows how experiment trackers fit into the 

Reviewer: ```suggestion
Here is an architecture diagram that shows how a model registry fits into the 
```

|--------------------------------------|----------|---------------|-------------------------------------------------------------------------------------------------|
| [MLflow](./mlflow.md)                | `mlflow` | `mlflow`      | Add MLflow as Model Registry to your stack                                                      | 
| [Custom Implementation](./custom.md) | _custom_ |               | _custom_                                                                                        | Extend the Model Registry abstraction and provide your own implementation |

If you would like to see the available flavors of Model Registry, you can 
use the command:

```shell
zenml model-registry flavor list
```

## How to use it

Model registries are an optional component in the ZenML stack that are tied to
the experiment tracker. This means that a model registry can only be used if you
are also using an experiment tracker. If you're not using an experiment tracker,
you can still store your models in ZenML, but you will need to manually retrieve
model artifacts from the artifact store. More information on this can be found

> in the [Fetching pipelines](../../starter-guide/pipelines/fetching-pipelines.md)
> documentation.

Reviewer: ```suggestion
in the [documentation on the post-execution workflow](../../starter-guide/pipelines/fetching-pipelines.md).
```
